mod-clans/src/main/java/gg/solarmc/loader/clans/ClanManager.java

```diff
@@ -0,0 +1,177 @@
+     * @param clan2 the second clan id, ally of the first
+     */
+    void invalidateAllianceCache(Integer clan1, Integer clan2) {
+        this.allianceCache.invalidate(clan1);
+        this.allianceCache.invalidate(clan2);
+    }
+
+    /**
+     * (a248 please review this method, i'm not sure if inserting two rows here is an issue
+     * because it will be called on clan generation for the first and for the second clan,
+     * and i'm not sure if this will lead to 2 rows {cache ignores duplicate} or to 4 rows { cache accepts duplicate} )
+     *
+     * Inserts 2 rows into the alliance cache, not order sensitive
+     * @param clan1 the first clan, ally of the second
+     * @param clan2 the second clan, ally of the first
+     */
+    void insertAllianceCache(Integer clan1, Integer clan2) {
+        this.allianceCache.put(clan1,clan2);
+        this.allianceCache.put(clan2,clan1);
+    }
+
+    /**
+     * Creates an empty clan with given name.
+     * @param name Name of the gg.solarmc.loader.clans.Clan to add
+     * @param transaction the tx
+     * @param player the to be owner of the clan
+     * @return created clan.
+     */
+    public Clan createClan(Transaction transaction, String name, SolarPlayer player) {
+        ClanDataObject owner = player.getData(ClansKey.INSTANCE);
+        ClansClanInfoRecord rec = transaction.getProperty(DSLContext.class)
+                .insertInto(CLANS_CLAN_INFO)
+                .columns(CLANS_CLAN_INFO.CLAN_NAME, CLANS_CLAN_INFO.CLAN_LEADER, CLANS_CLAN_INFO.CLAN_KILLS, CLANS_CLAN_INFO.CLAN_DEATHS, CLANS_CLAN_INFO.CLAN_ASSISTS)
+                .values(name, owner.getUserId(), 0, 0, 0)
+                .returning()
+                .fetchOne();
+
+        if (rec == null) throw new IllegalStateException("Failed to insert new gg.solarmc.loader.clans.Clan by name " + name);
+
+        ClanMember ownerAsMember = owner.asClanMember(transaction);
+        Set<ClanMember> memberSet = new HashSet<>();
+        memberSet.add(ownerAsMember);
+
+        Clan returned = new Clan(rec.getClanId(),rec.getClanName(),rec.getClanKills(),rec.getClanDeaths(),rec.getClanAssists(),this,memberSet,ownerAsMember);
+
+        clans.put(returned.getID(),returned);
+
+        return returned;
+    }
+
+    /**
+     * Deletes a clan from the cache and table, and deletes alliances
+     * if the clan is in an alliance.
+     *
+     * @param transaction the tx
+     * @param clan the clan to delete
+     * @throws IllegalStateException if the clan does not exist in the table.
+     */
+    public void deleteClan(Transaction transaction, Clan clan) {
+        int i = transaction.getProperty(DSLContext.class)
+                .delete(CLANS_CLAN_INFO)
+                .where(CLANS_CLAN_INFO.CLAN_ID.eq(clan.getID()))
+                .execute();
+
+        if (i == 1) {
+            throw new IllegalStateException("Clan does not exist in table!");
+        }
+
+        clan.getAlliedClan(transaction).ifPresent(ally -> {
```

Comment:
Don't perform a database query merely to uncache in-memory data. You can improve this